src/macros.jl

```diff
@@ -359,23 +368,42 @@ macro wrap(ex)
 
     fullargs = ex.args[2:length(ex.args)]
 
-    # strip type annotation from arguments for use in the call to GAP
-    args = [x isa Symbol ? x : x.args[1] for x in fullargs]
-
+    # splits the arguments with type annotations into expressions for the lhs and rhs
+    # of the call of the form `func(args) = GAP.Globals.func(args)`:
+    # - type annotations have to be removed for the rhs
+    # - if the type annotation is `GapObj`, the rhs has to be wrapped in `GAP.GapObj(...)` and the lhs has no type annotation
+    tempargs = [
+        begin
+            if x isa Symbol
+                (x, x)
+            elseif x.head == :(::) && length(x.args) == 2
+                var = x.args[1]
+                typeannot = x.args[2]
+                if typeannot in [:GapObj, :(GAP.GapObj)]
+                    (var, :(GAP.GapObj($var)))
```

Comment:
```suggestion
                if typeannot in [:GapObj, :(GAP.GapObj)]
                    (var, :(GAP.GapObj($var)))
                elseif typeannot == :(GAP.Obj)
                    (var, :(GAP.Obj($var)))
                elseif typeannot in [:GapInt, :(GAP.GapInt)]
                    (var, :(GAP.GapInt($var)))
```

Reply:
Like this, @ThomasBreuer @lgoettgens ?

Reply:
one a purely syntactical level this looks right. But I don't really know about the capabilities of the `GAP.Obj(...)` and `GAP.GapInt(...)` constructors to judge this

Reply:
Concerning the possible "restriction" of functions to `::GapObj` arguments, this is what I wanted to express:
On the one hand, GAP's `IsFreeGroup` is a filter that accepts any `GAP.Obj`, hence `::(GAP.Obj)` would be correct.
On the other hand, we want to call `IsFreeGroup` from Oscar only with a `::GapObj` argument, hence it is o.k. to define it like this in the `@wrap` call.

Concerning automatic indirection from Oscar objects to GAP objects via `GAP.Obj`:
I see that this works already for Oscar groups because Oscar defines `GAP.julia_to_gap(obj::GAPGroup) = obj.X` and not just `GapObj(obj::GAPGroup) = obj.X`.
(Thus I did this the wrong way for `GAPGroupCharacterTable` ...)

I think this new feature looks elegant, but understanding the Oscar code becomes more subtle:
We have to distinguish carefully between the use of `GapObj` for the *conversion* of objects (which creates a new object in each call and should be avoided if possible) and its use for *indirection* to objects which already exist (which is fine).

Reply:
@lgoettgens can you also adapt this suggestion?

Reply:
Done